### ISSUE TITLE
Page Title is always overwritten on navigation

### ISSUE THREAD

**Issue by markuski** [vaadin/flow#18017]
### Description of the bug

On navigation, the page title is always overwritten. 
When no PageTitle Annotation and no HasDynamicTitle is found, then a navigation always overwrites the current page title with an empty string.

### Expected behavior

When there is no PageTitle Annotation and the component does not implement HasDynamicTitle,
then the current page title should not be overwritten.

### Minimal reproducible example

call ui.getPage().setTitle(titleBeforeNavigation);
-> navigate to component without title (no Page Title Annotation and no HasDynamicTitle)
-> Title is reset to ""

possible fix: https://github.com/vaadin/flow/commit/4d97f50c330fcc53b53ebddb0301c316672504f9

### Versions

- Vaadin / Flow version: 24.2.2
- Java version: 17

**Comment by knoobie** [comment]
I would argue that every page has to define a title - see [WCAG 2.4.2 Criteria](https://www.w3.org/WAI/WCAG22/Understanding/page-titled.html). Therefore removing the page title, if not present, is a good default to fail accessibility audits to identify those pages and fix them.

**Comment by jcgueriaud1** [comment]
As @knoobie said, I don't see why the title of the previous page is correct. It's very likely a wrong title for the page, if it's empty it's easier to detect and fix it.

**Comment by markuski** [comment]
Then Page.setTitle() barely makes any sense, if it's going to get overwritten anyway.

**Comment by jcgueriaud1** [comment]
Ok, I think I understand. I'm usually calling the Page.setTitle after the navigation (since I'm using the code when I need data from the database).

```
@Route("test")
public class TestView extends Div implements BeforeEnterObserver {

    public TestView() {
        TextField textField = new TextField();
        add(textField);
    }


    @Override
    public void beforeEnter(BeforeEnterEvent event) {
        event.getUI().getPage().setTitle("My title");
    }
}
```

In that example, the page title is empty but should be filled with "My title"

**Comment by markuski** [comment]
I wanted to set the title for any route in one place within a afterNavigationListener.
This solution would ensure consistency, as my title consists of static and dynamic parts, something like "DEV - My Route"

**Comment by knoobie** [comment]
That's possible and the default use-case.

**Comment by jcgueriaud1** [comment]
> I wanted to set the title for any route in one place within a afterNavigationListener.
This solution would ensure consistency, as my title consists of static and dynamic parts, something like "DEV - My Route"

In your case, why don't you use `HasDynamicTitle`? (I'm not saying that the issue is invalid and `ui.getPage().setTitle()` is confusing)

**Comment by markuski** [comment]
because for every route i have to do the same thing and inject the same data to generate the title, it would be my preference to set the title in one place.

**Comment by markuski** [comment]
@knoobie its not possible now, because afterNavigation is fired before updatePageTitle.
`        fireAfterNavigationListeners(
                new AfterNavigationEvent(locationChangeEvent, parameters),
                afterNavigationHandlers);

        updatePageTitle(event, componentInstance);`
        
Maybe Vaadin should fire the afterNavigation event as the last instruction of AbstractNavigationStateRenderer.handle ?

**Comment by mshabarov** [comment]
The problem is because title is update after "afterNavigation" event if fired, see
```java
        fireAfterNavigationListeners(
                new AfterNavigationEvent(locationChangeEvent, parameters),
                afterNavigationHandlers);

        updatePageTitle(event, componentInstance);
```
So your changes in `afterNavigationListener` are replaced one line later with the page title which is empty.
We had a fix in the past that changed the order of these events, see https://github.com/vaadin/flow/pull/13825.

**Comment by knoobie** [comment]
Interesting good to know, I had expected the same, but changing this behavior could also lead to multiple people's code breaking. 

I'm personally using this and implementing this on my views: 

```java

public interface HasPageTitle extends HasDynamicTitle {

  String getRouteName();

  @Override
  default String getPageTitle() {
    return getRouteName() + " | " + " My App";
  }
}

```

**Comment by vojtapol** [comment]
I disagree with the responses here. I have an app that is very dynamic and for whatever reason, I don't care about the page titles at all. I just want the page title to be set in one place and say "My Amazing App". This is something I could easily do in a PWA app like Vue.js where there is only one `<title>` tag in the `index.html` file.

In Vaadin, I set `@PageTitle` on the `AppShellConfigurator` and the page title is retained during navigation. To be specific, it gets set to empty string but then immediately overwritten back in `UI#adjustPageTitle`. The only time this call does not happen is when the navigation target has `QueryParameters`. In that case `adjustPageTitle()` is never called and the `<title>` element is set to empty string.

**Comment by mshabarov** [comment]
@vojtapol fair point, thanks!
`AppShellConfigurator` has indeed an API for setting initial page title, with the annotation or via `AppShellSettings` that is applicable to any routes that have to explicit title defined.
I'd treat this issue as a bug, because of 

> The only time this call does not happen is when the navigation target has QueryParameters
